src/nunit.analyzers/Helpers/DiagnosticsHelper.cs

```diff
@@ -0,0 +1,26 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Operations;
+using NUnit.Analyzers.Constants;
+
+namespace NUnit.Analyzers.Helpers
+{
+    internal static class DiagnosticsHelper
+    {
+        public static bool LastArgumentIsNonParamsArray(ImmutableArray<IArgumentOperation> arguments)
```

Comment:
This seems like an unnecessarily strong requirement. What we want is [a read-only collection of elements that can be accessed by index](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlylist-1):
```suggestion
        public static bool LastArgumentIsNonParamsArray(IReadOnlyList<IArgumentOperation> arguments)
```
We can keep other parts of the code because [`ImmutableArray<T>` implements `IReadOnlyList<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1).

Reply:
On second thought, we can go a step further and just do IReadOnlyCollection<T> and .Last() below.

Reply:
Yes we could, but all calls are from direct `IInvocationOperation.Arguments` which uses an `ImmutableArray`:

![image](https://github.com/nunit/nunit.analyzers/assets/8655218/28ef0c30-7396-4489-ba33-a9a34d8ff339)

If there would have been code that had explicitly create the `ImmutableArray` I would agree.

Calling Linq methods on indexable collections raised CA1826:

![image](https://github.com/nunit/nunit.analyzers/assets/8655218/a812801f-977a-44ff-b7ce-f02284c9e8f1)

We do not know IF Linq has a short cut for collections.

Reply:
One more item. As all source arguments are `ImmutableArray` which is a `struct`. Declaring the parameter as an `interface` will cause expensive [boxing](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/types/boxing-and-unboxing#performance).

This can be prevented by making the method generic:
```csharp
public static bool LastArgumentIsNonParamsArray<T>(T arguments)
    where T : IReadOnlyList<IArgumentOperation>
```

In addition calling properties/method on an interface calls through a virtual method table.
For performance reasons .NET 8 has a new [rule](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1859) asking you to make then concrete where possible.
There is a table in [performance-improvements-in-net-8](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/) about the difference of using `IList<T>` vs `List<T>`.